### DISCUSSION TITLE
Help with #1149 (different icons for different modi)

### DISCUSSION THREAD

**Discussion by pedrolimasi** [davatorium/rofi#1535] [Q&A] [answered]
In the issue #1149, @DaveDavenport stated that the reported bug is actually the intended behaviour and the issue was closed.
Although I was not the one who opened the issue, I ran into it with myself and was wondering if anyone could point me in the right direction.

The app I'm running has no icon in window modi. No solution to the problem was suggested, and while digging through articles on Xorg's `NET_WM_ICON` property,  I have come up with nothing that helped. I understand the solution would not come through rofi's side, but if I could solve it myself in my machine, that's good enough for me.

All help is appreciated, thanks.

**Comment by DaveDavenport** [comment, marked as answer]
Its not really intended behaviour as I would like it to be better :-/, it is just how things work according to spec.

the best solution is for the application to set an icon.
(you can run xprop from a terminal and click on the application to see what it sets)

The fallback is that the class name resolves an icon in the icon theme.
If the application sets a class name, you can try adding an image with that name.

**Reply by pedrolimasi** [reply]
Thanks for the quick reply Dave. 
`xprop` tells me the icon is `(not shown)`
The class is properly setup, though.
Got any recomendations on how to persistently set an icon for a `WM_CLASS`?

**Reply by DaveDavenport** [reply]
can you pastebin the xprop output?

**Reply by DaveDavenport** [reply]
(what rofi version?, the class part (and a fix to it) is only in latest)

**Reply by pedrolimasi** [reply]
It's an Electron wrapper for the Notion enhanced ([repo link](https://github.com/notion-enhancer/notion-repackaged))

<details><summary> xprop output</summary>

```
_NET_WM_ICON_GEOMETRY(CARDINAL) = 1120, 1080, 0, 0
WM_STATE(WM_STATE):
		window state: Normal
		icon window: 0x0
_NET_WM_DESKTOP(CARDINAL) = 0
_NET_WM_ALLOWED_ACTIONS(ATOM) = _NET_WM_ACTION_CHANGE_DESKTOP, _NET_WM_ACTION_CLOSE, _NET_WM_ACTION_MOVE, _NET_WM_ACTION_MINIMIZE, _NET_WM_ACTION_RESIZE, _NET_WM_ACTION_FULLSCREEN, _NET_WM_ACTION_MAXIMIZE_HORZ, _NET_WM_ACTION_MAXIMIZE_VERT, _NET_WM_ACTION_ABOVE, _NET_WM_ACTION_BELOW, _OB_WM_ACTION_UNDECORATE
_KDE_NET_WM_FRAME_STRUT(CARDINAL) = 0, 0, 0, 0
_NET_FRAME_EXTENTS(CARDINAL) = 0, 0, 0, 0
_NET_WM_STATE(ATOM) = _OB_WM_STATE_UNDECORATED
_NET_WM_ICON(CARDINAL) = 	Icon (48 x 48):
	(not shown)

_OB_APP_TYPE(UTF8_STRING) = "normal"
_OB_APP_TITLE(UTF8_STRING) = "Notion Desktop"
_OB_APP_GROUP_CLASS(UTF8_STRING) =
_OB_APP_GROUP_NAME(UTF8_STRING) =
_OB_APP_CLASS(UTF8_STRING) = "notion-app-enhanced"
_OB_APP_NAME(UTF8_STRING) = "notion-app-enhanced"
_OB_APP_ROLE(UTF8_STRING) = "browser-window"
_NET_WM_VISIBLE_ICON_NAME(UTF8_STRING) = "Linux"
_NET_WM_VISIBLE_NAME(UTF8_STRING) = "Linux"
_NET_WM_USER_TIME(CARDINAL) = 1412409
WM_NAME(UTF8_STRING) = "Linux"
_NET_WM_NAME(UTF8_STRING) = "Linux"
WM_NORMAL_HINTS(WM_SIZE_HINTS):
		program specified location: 541, 99
		program specified maximum size: 2147483647 by 2147483647
_MOTIF_WM_HINTS(_MOTIF_WM_HINTS) = 0x2, 0x0, 0x0, 0x0, 0x0
XdndAware(ATOM) = BITMAP
_NET_WM_BYPASS_COMPOSITOR(CARDINAL) = 2
_GTK_HIDE_TITLEBAR_WHEN_MAXIMIZED(CARDINAL) = 1
WM_WINDOW_ROLE(STRING) = "browser-window"
WM_CLASS(STRING) = "notion-app-enhanced", "notion-app-enhanced"
_NET_WM_WINDOW_TYPE(ATOM) = _NET_WM_WINDOW_TYPE_NORMAL
_NET_WM_PID(CARDINAL) = 2581
WM_CLIENT_MACHINE(STRING) = "arch"
WM_PROTOCOLS(ATOM): protocols  WM_DELETE_WINDOW, _NET_WM_PING, _NET_WM_SYNC_REQUEST
```

</details>

**Reply by pedrolimasi** [reply]
`rofi -v`
`Version: 1.7.0`

**Reply by DaveDavenport** [reply]
curious it has an icon, but apparently one neither rofi or xprop can parse. 
if that is the case, I might be able to support it.

**Reply by DaveDavenport** [reply]
installed it, I am not getting an icon entry in xprop.  this is fun.

**Reply by DaveDavenport** [reply]
If you have rofi 1.7.1  you can add an image with the name `notion-app-enhanced` to your icon theme and that should then pick-up.

**Reply by DaveDavenport** [reply]
![rofi-2021-11-29-2255-00000](https://user-images.githubusercontent.com/406245/143949031-c87983e4-f44a-4bd7-a837-1a6e4bef45aa.png)

I think if you upgrade rofi, it might just work out of the box if you have it installed from a  deb/rpm file. (and not using appimage)

**Reply by pedrolimasi** [reply]
Great! Thanks a lot man. I'll wait for the AUR to update, then test it. If it worked on your side, I'll mark this thread as answered.

**Reply by DaveDavenport** [reply]
please check the release notes for 1.7.1 because of a bug in older version a small config file tweak might be needed.

**Comment by pedrolimasi** [comment]
Apparently it's a thing isn't it? Applications failing to set their icon. Even found a specialized tool https://github.com/xeyownt/xseticon.

**Reply by DaveDavenport** [reply]
did not know that tool.
